### PR TITLE
Repos and Forks can not add multiple domains for the same key

### PR DESCRIPTION
**Pull request by nhunzaker**
This commit removes the ability to add multiple domains for the same key. This allows us to:

1. Remove the meta-domain
2. Remove the ADD_DOMAIN lifecycle action
3. Avoid merging data to avoid "spilling" keys

**Breaking changes**

1. Microcosms can no longer implement `register()`. This was intended to make it easier to build a Microcosm, but it is never used in real-world apps™️ and is problematic for future work.
2. Forks can not add domains on top of parents
3. Microcosms can not add multiple domains for the same key

Fixes https://github.com/vigetlabs/microcosm/issues/427.

### CODE REVIEW

File: packages/microcosm/src/domain-engine.js [hunk 1]
```diff
@@ -76,28 +69,50 @@ class DomainEngine {
...
+    this.lifecycle[PATCH.toString()].push({
+      key: key,
+      scope: domain,
+      local: true,
+      steps: [
+        (state: *, data: any) => {
+          return data[key] !== undefined ? data[key] : state
+        }
+      ]
+    })
```
Comment (efatsi) [inline packages/microcosm/src/domain-engine.js]: can you explain what's up with `RESET.toString()` and `PATCH.toString()`? And what these two pushes do?

File: packages/microcosm/src/domain-engine.js [hunk 2]
```diff
@@ -112,60 +127,78 @@ class DomainEngine {
...
-      } else {
-        // No: use the existing snapshot value (memoizing the domain handler)
-        result = set(result, key, get(snapshot.next, key))
+        // Recalculate state from the last answer
+        next = reduce(steps, action.payload, last, scope)
+      }
+
+      if (answer === state && next !== answer[key]) {
+        answer = clone(state)
       }
```
Comment (efatsi) [inline packages/microcosm/src/domain-engine.js]: can you summarize what this conditional is checking, and then doing if it returns true?

Comment (nhunzaker) [reply packages/microcosm/src/domain-engine.js]: Yep! This check is saying "If the answer object reference is the same as the state object reference, and the value has changed, then clone the answer".

This is pretty low level. Basically I want to avoid shallow cloning the next state object every time a domain handler runs.

File: packages/microcosm/src/domain-engine.js (hunk 1 above)
Comment (nhunzaker) [reply packages/microcosm/src/domain-engine.js]: Yep. 

`PATCH.toString()` is to tell Flow that the underlying value is a string. This is _lame_.

These two pushes replace the MetaDomain as the method of modifying data after `repo.reset` or `repo.patch` fire.

For reset:

Update the value for the key the domain is managing if the key is defined in the data payload. Otherwise fall back to the initial state. So `repo.reset()` reverts to initial state, but `repo({ key: 'value' })` sets a specific value.

For patch:

The same is true, however the fallback is the _current_ value, rather than the initial state.
